File: golang/solution/solution153.go
package solution

import "fmt"

// 搜索旋转数组最小值
func findMin(nums []int) int {
	if len(nums) == 0 {
		panic("nums can not be empty")
	} else if len(nums) == 1 {
		return nums[0]
	} else if len(nums) == 2 {
		return func(i, j int) int {
			if i < j {
				return i
			}
			return j
		}(nums[0], nums[1])
	} else {
		left, right := 0, len(nums)-1
		for left < right {
			mid := (left + right) / 2
			if nums[mid] < nums[right] {
				// 从[mid, right]是递增序列，最小值在 mid左边，收缩右边界
				right = mid
			} else {
				// 收缩左边界
				left = mid + 1
			}
		}
		return nums[left]
	}
}

func Solution153() {
	var nums []int = []int{4, 5, 6, 0, 1, 2, 3}
	res := findMin(nums)
	fmt.Println(res)
}

File: leetcode_alg/solution153.cpp
//
// Created by wangheng on 1/10/24.
//

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int findMin(vector<int> &nums) {
        int left = 0;
        int right = nums.size() - 1;
        int mid;
        while (left < right) {
            mid = left + (right - left) / 2;
            if (nums[mid] > nums[right]) {
                // 中间值大于右边值，最小值在右边，收缩左边界
                left = mid + 1;
            } else {
                // 中间值小于右边值，最小值在左边，收缩右边界
                right = mid;
            }
        }
        return nums[left];
    }
};

int main(int argc, char *argv[]) {
    auto s = new Solution;
    vector<int> nums{5, 1, 2, 3, 4};
    cout << "结果是:" << s->findMin(nums) << endl;
    delete s;
    return 0;
}